apiextensions/v1alpha1/types_compatibilityrequirement.go

```diff
@@ -0,0 +1,391 @@
+	// +kubebuilder:validation:MaxItems=255
+	// +listType=set
+	// +optional
+	Additional []APIVersionString `json:"additional,omitempty"`
+}
+
+// APIExcludedField describes a field in the schema which will not be validated by
+// crdSchemaValidation or objectSchemaValidation.
+type APIExcludedField struct {
+	// path is the path to the field in the schema.
+	// Paths are dot-separated field names (e.g., "fieldA.fieldB.fieldC") representing nested object fields.
+	// Each field name must be a valid Kubernetes CRD field name: start with a letter, contain only
+	// letters, digits, and underscores, and be between 1 and 63 characters in length.
+	// A path may contain at most 32 fields.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=1024
+	// +kubebuilder:validation:XValidation:rule="self.split('.', 32).all(f, f.matches('^[a-zA-Z][a-zA-Z0-9_]{0,63}$'))",message="path must be dot-separated field names, each starting with a letter and containing only letters, digits, and underscores not exceeding 63 characters. There may be at most 32 fields in the path."
+	// +required
+	Path string `json:"path,omitempty"`
+
+	// version is the version of the API that the field is excluded from.
+	// When not specified, the field is excluded from all versions.
+	// When present, must be a valid Kubernetes API version string, with a
+	// maximum length of 255 characters.
+	// +optional
+	Version APIVersionString `json:"version,omitempty"`
+}
+
+// CompatibilitySchema defines the schema used by crdSchemaValidation and objectSchemaValidation.
+type CompatibilitySchema struct {
+	// customResourceDefinition contains the complete definition of the CRD for schema and object validation purposes.
+	// This field is required.
+	// +required
+	CustomResourceDefinition CRDData `json:"customResourceDefinition,omitzero"`
+
+	// requiredVersions specifies a subset of the CRD's API versions which will be asserted for compatibility.
+	// This field is required.
+	// +required
+	RequiredVersions APIVersions `json:"requiredVersions,omitzero"`
+
+	// excludedFields is a set of fields in the schema which will not be validated by
+	// crdSchemaValidation or objectSchemaValidation.
+	// The list may contain at most 64 fields.
+	// When not specified, all fields in the schema will be validated.
+	// +kubebuilder:validation:MaxItems=64
+	// +listType=atomic
+	// +optional
+	ExcludedFields []APIExcludedField `json:"excludedFields,omitempty"`
+}
+
+// CustomResourceDefinitionSchemaValidation ensures that updates to the installed CRD are compatible with this compatibility requirement.
+type CustomResourceDefinitionSchemaValidation struct {
+	// action determines whether violations are rejected (Deny) or admitted with an API warning (Warn).
+	// Valid options are:
+	//   Deny - incompatible CRDs will be rejected and not admitted to the cluster.
+	//   Warn - incompatible CRDs will be allowed but a warning will be generated in the API response.
```

Comment:
Normally formatted as 

```suggestion
	// Valid values are Deny and Warn.
	// When set to Deny, incompatible CRDs will be rejected and not admitted to the cluster.
	// When set to Warn, incompatible CRDs will be allowed but a warning will be generated in the API response.
```